File: javascript/json-transform/src/functions/TransformerFunctionMerge.ts
import TransformerFunction from "./common/TransformerFunction";
import { ArgType } from "./common/ArgType";
import FunctionContext from "./common/FunctionContext";
import { JsonMergeOptions, merge } from "../JsonHelpers";
import { AsyncSequence, asyncSequenceOf } from "@wortise/sequency";

class TransformerFunctionMerge extends TransformerFunction {
  constructor() {
    super({
      arguments: {
        deep: { type: ArgType.Boolean, position: 1, defaultBoolean: false },
        arrays: { type: ArgType.Boolean, position: 2, defaultBoolean: false },
      },
    });
  }

  override async apply(context: FunctionContext): Promise<any> {
    const streamer = await context.getJsonElementStreamer(null);
    if (streamer == null || streamer.knownAsEmpty()) return null;

    const deep = await context.getBoolean("deep");
    const arrays = await context.getBoolean("arrays");

    const options: JsonMergeOptions = {
      deep: deep ?? undefined,
      concatArrays: arrays ?? undefined,
    };

    let result: Record<string, any> = {};

    return (asyncSequenceOf(asyncSequenceOf(result), streamer.stream()).flatten() as AsyncSequence<any>).reduce(
      async (acc, value) => {
        return merge(result, value, options);
      },
    );
  }
}

export default TransformerFunctionMerge;

File: java/json-transform/src/main/java/co/nlighten/jsontransform/functions/TransformerFunctionMerge.java
package co.nlighten.jsontransform.functions;

import co.nlighten.jsontransform.adapters.JsonAdapter;
import co.nlighten.jsontransform.functions.common.*;

import java.util.Map;
import java.util.stream.Stream;

public class TransformerFunctionMerge extends TransformerFunction {

    public TransformerFunctionMerge() {
        super(FunctionDescription.of(
            Map.of(
                "deep", ArgumentType.of(ArgType.Boolean).position(0).defaultBoolean(false),
                "arrays", ArgumentType.of(ArgType.Boolean).position(1).defaultBoolean(false)
            )
        ));
    }
    @Override
    public Object apply(FunctionContext context) {
        var streamer = context.getJsonElementStreamer(null);
        if (streamer == null || streamer.knownAsEmpty())
            return null;

        var deep = context.getBoolean("deep");
        var arrays = context.getBoolean("arrays");

        var options = new JsonAdapter.JsonMergeOptions(deep, arrays);

        var adapter = context.getAdapter();
        var result = adapter.createObject();

        return ((Stream<Object>)streamer.stream())
                .reduce(result, (acc, value) -> adapter.merge(acc, value, options));
    }
}
